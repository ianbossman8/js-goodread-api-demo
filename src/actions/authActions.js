export const ACTION_CONSTANTS = {
  AUTH: 'AUTH',
  NOT_AUTH: 'NOT_AUTH',
}

export function getAuthAction(type) {
  return {
    type,
  }
}
