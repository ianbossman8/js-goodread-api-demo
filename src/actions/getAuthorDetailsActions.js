import getAuthorFollowers, { actions as authorActions } from './getAuthorFollowersActions'

export const ACTION_CONSTANTS = {
  AUTHOR_DETAILS_FETCHING: 'AUTHOR_DETAILS_FETCHING',
  AUTHOR_DETAILS_RECEIVED: 'AUTHOR_DETAILS_RECEIVED',
  AUTHOR_DETAILS_ERROR: 'AUTHOR_DETAILS_ERROR',
}

export const actions = {
  receivedAuthorDetails: function (result) {
    return {
      type: ACTION_CONSTANTS.AUTHOR_DETAILS_RECEIVED,
      result,
    }
  },
  fetchingAuthorDetails: function () {
    return {
      type: ACTION_CONSTANTS.AUTHOR_DETAILS_FETCHING,
    }
  },
  errorAuthorDetails: function (error) {
    return {
      type: ACTION_CONSTANTS.AUTHOR_DETAILS_ERROR,
      error,
    }
  },
}

export default function getAuthorDetails(authodId) {
  return async (dispatch, getState, { netlifyEndpoint }) => {
    try {
      dispatch(actions.fetchingAuthorDetails())
      dispatch(authorActions.resetAuthorFollowers())
      const authorDetailUrl = new URL(`/.netlify/functions/author/${authodId}`, netlifyEndpoint)

      const response = await fetch(authorDetailUrl)

      if (response.ok) {
        const parsedRes = await response.json()

        if (getState().auth.isAuth && typeof parsedRes.GoodreadsResponse.author.user !== 'undefined') {
          dispatch(getAuthorFollowers(parsedRes.GoodreadsResponse.author.user.id))
        }

        return dispatch(actions.receivedAuthorDetails(parsedRes.GoodreadsResponse.author))
      }
      throw new Error('author detail api error')
    } catch (error) {
      dispatch(actions.errorAuthorDetails(error.message))
    }
  }
}
