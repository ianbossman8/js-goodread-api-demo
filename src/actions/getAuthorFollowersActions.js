export const ACTION_CONSTANTS = {
  AUTHOR_FOLLOWERS_FETCHING: 'AUTHOR_FOLLOWERS_FETCHING',
  AUTHOR_FOLLOWERS_RECEIVED: 'AUTHOR_FOLLOWERS_RECEIVED',
  AUTHOR_FOLLOWERS_ERROR: 'AUTHOR_FOLLOWERS_ERROR',
  AUTHOR_FOLLOWERS_RESET: 'AUTHOR_FOLLOWERS_RESET',
}

export const actions = {
  receivedAuthorFollowers: function (result) {
    return {
      type: ACTION_CONSTANTS.AUTHOR_FOLLOWERS_RECEIVED,
      result,
    }
  },
  fetchingAuthorFollowers: function () {
    return {
      type: ACTION_CONSTANTS.AUTHOR_FOLLOWERS_FETCHING,
    }
  },
  errorAuthorFollowers: function (error) {
    return {
      type: ACTION_CONSTANTS.AUTHOR_FOLLOWERS_ERROR,
      error,
    }
  },
  resetAuthorFollowers: function () {
    return {
      type: ACTION_CONSTANTS.AUTHOR_FOLLOWERS_RESET,
    }
  },
}

export default function getAuthorFollowers(authorUserId, page = 1) {
  return async (dispatch, _, { netlifyEndpoint }) => {
    try {
      dispatch(actions.fetchingAuthorFollowers())

      const authorFollowersUrl = new URL(`/.netlify/functions/author-followers/${authorUserId}`, netlifyEndpoint)
      const params = {
        page,
      }
      authorFollowersUrl.search = new URLSearchParams(params)

      const response = await fetch(authorFollowersUrl, {
        method: 'GET',
        Accept: 'application/json',
        credentials: 'include',
      })

      if (response.ok) {
        const parsedRes = await response.json()

        return dispatch(actions.receivedAuthorFollowers(parsedRes.GoodreadsResponse.followers.user))
      }

      throw new Error('followers api error')
    } catch (error) {
      dispatch(actions.errorAuthorFollowers(error.message))
    }
  }
}
