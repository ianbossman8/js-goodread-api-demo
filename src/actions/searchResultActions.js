export const ACTION_CONSTANTS = {
  SEARCH_RESULT_FETCHING: 'SEARCH_RESULT_FETCHING',
  SEARCH_RESULT_RECEIVED: 'SEARCH_RESULT_RECEIVED',
  SEARCH_RESULT_ERROR: 'SEARCH_RESULT_ERROR',
}

export const actions = {
  receivedSearchResults: function (result) {
    return {
      type: ACTION_CONSTANTS.SEARCH_RESULT_RECEIVED,
      result,
    }
  },
  fetchingSearchResult: function () {
    return {
      type: ACTION_CONSTANTS.SEARCH_RESULT_FETCHING,
    }
  },
  errorSearchResult: function (error) {
    return {
      type: ACTION_CONSTANTS.SEARCH_RESULT_ERROR,
      error,
    }
  },
}

export default function getSearchResults(query, search = 'all', page = 1) {
  return function (signal) {
    return async (dispatch, _, { netlifyEndpoint }) => {
      dispatch(actions.fetchingSearchResult())
      try {
        const goodreadSearchUrl = new URL('/.netlify/functions/search', netlifyEndpoint)

        const params = {
          q: query,
          page,
          search,
        }

        goodreadSearchUrl.search = new URLSearchParams(params)

        const response = await fetch(goodreadSearchUrl, { signal })

        if (response.ok) {
          const parsedRes = await response.json()

          return dispatch(actions.receivedSearchResults(parsedRes.GoodreadsResponse.search))
        }

        throw new Error('search api error')
      } catch (error) {
        dispatch(actions.errorSearchResult(error.message))
      }
    }
  }
}
