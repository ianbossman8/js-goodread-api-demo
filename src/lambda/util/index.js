import OAuth from 'oauth'

export default new OAuth.OAuth(
  'https://www.goodreads.com/oauth/request_token',
  'https://www.goodreads.com/oauth/access_token',
  process.env.API_KEY,
  process.env.API_SECRET,
  '1.0A',
  process.env.NODE_ENV !== 'development'
    ? 'http://localhost:9000/auth-callback'
    : 'https://infallible-wilson-4e1a09.netlify.app/auth-callback',
  'HMAC-SHA1'
)
