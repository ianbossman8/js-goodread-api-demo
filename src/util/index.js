import OAuth from 'oauth'

function getOAuthReqCode() {
  const oauth1 = new OAuth.OAuth(
    'https://www.goodreads.com/oauth/request_token',
    'https://www.goodreads.com/oauth/access_token',
    process.env.API_KEY,
    process.env.API_SECRET,
    '1.0A',
    'http://localhost:9000/auth-callback',
    'HMAC-SHA1'
  )

  if (
    typeof process.env.GOODREAD_REQ_TOKEN === 'undefined' &&
    typeof process.env.GOODREAD_REQ_TOKEN_SECRET === 'undefined'
  ) {
    oauth1.getOAuthRequestToken(function (err, token, token_secret) {
      if (!err) {
        process.env.GOODREAD_REQ_TOKEN = token
        process.env.GOODREAD_REQ_TOKEN_SECRET = token_secret
      }
    })
  }

  return oauth1
}

export default getOAuthReqCode()
