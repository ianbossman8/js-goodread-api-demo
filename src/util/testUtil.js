import React from 'react'
import { Provider } from 'react-redux'

export function AppWrapper(props) {
  return <Provider store={props.store}>{props.children}</Provider>
}
