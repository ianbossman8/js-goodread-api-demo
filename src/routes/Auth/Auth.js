import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Link, useLocation } from 'react-router-dom'
import { getAuthAction, ACTION_CONSTANTS } from '../../actions/authActions'

function Auth() {
  const location = useLocation()
  const dispatch = useDispatch()

  const authCode = location.search[location.search.length - 1]

  useEffect(() => {
    if (authCode !== '0') {
      sessionStorage.setItem('isAuth', true)
      dispatch(getAuthAction(ACTION_CONSTANTS.AUTH))
    }
  })

  return (
    <div className="auth-page">
      <p>{authCode === '0' ? 'not auth' : 'auth'}</p>
      <Link to="/">Link to homepage</Link>
    </div>
  )
}

export default Auth
