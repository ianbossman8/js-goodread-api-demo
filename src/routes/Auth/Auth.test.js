import React from 'react'
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils'
import configureStore from 'redux-mock-store'
import { render, screen } from '@testing-library/react'
import { AppWrapper } from '../../util/testUtil'
import { ACTION_CONSTANTS } from '../../actions/authActions'
import Auth from './Auth'

describe('Auth component', () => {
  const store = configureStore()({})

  afterEach(() => {
    store.clearActions()
  })

  const wrappedComponent = (route = '/auth?auth=0') => (
    <AppWrapper store={store}>
      <MemoryRouter initialEntries={[route]}>
        <Auth />
      </MemoryRouter>
    </AppWrapper>
  )

  it('should match snapshot', () => {
    const { container } = render(wrappedComponent())

    expect(container.firstChild).toMatchSnapshot()
  })

  describe('ui display', () => {
    it('should have correct display with auth=0', () => {
      render(wrappedComponent())
      expect(screen.getByRole('link').textContent).toBe('Link to homepage')
      expect(screen.getByText(/not auth/i)).toBeInTheDocument()
    })

    it('should have correct display with auth=1', () => {
      render(wrappedComponent('/auth?auth=1'))
      expect(screen.getByRole('link').textContent).toBe('Link to homepage')
      expect(screen.getByText(/auth/i)).toBeInTheDocument()
    })
  })

  describe('page change', () => {
    let testHistory, testLocation
    render(
      <AppWrapper store={store}>
        <MemoryRouter initialEntries={['/auth?auth=1']}>
          <Auth />
          <Route
            path="*"
            render={({ history, location }) => {
              testHistory = history
              testLocation = location
              return null
            }}
          />
        </MemoryRouter>
      </AppWrapper>
    )

    expect(testLocation.pathname).toBe('/auth')

    act(() => {
      const goHomeLink = document.querySelector('a')
      goHomeLink.dispatchEvent(new MouseEvent('click', { bubbles: true }))
    })

    expect(testLocation.pathname).toBe('/')
  })

  describe('dispatch', () => {
    it('should dispatch action AUTH', () => {
      render(wrappedComponent('/auth?auth=1'))

      expect(store.getActions()[0].type).toBe(ACTION_CONSTANTS.AUTH)
    })
  })
})
