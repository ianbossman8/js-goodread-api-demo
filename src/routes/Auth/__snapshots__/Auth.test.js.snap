// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth component should match snapshot 1`] = `
<div
  class="auth-page"
>
  <p>
    not auth
  </p>
  <a
    href="/"
  >
    Link to homepage
  </a>
</div>
`;
